crates/ide-assists/src/handlers/generate_trait_from_impl.rs

```diff
@@ -0,0 +1,337 @@
+//     const CONST_ASSOC: usize = N * 4;
+//
+//     fn create() -> Option<()>;
+//
+//     const_maker! {i32, 7}
+// }
+//
+// impl<const N: usize> NewTrait<N> for Foo<N> {
+//     // Used as an associated constant.
+//     const CONST_ASSOC: usize = N * 4;
+//
+//     fn create() -> Option<()> {
+//         Some(())
+//     }
+//
+//     const_maker! {i32, 7}
+// }
+// ```
+pub(crate) fn generate_trait_from_impl(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    // Get AST Node
+    let impl_ast = ctx.find_node_at_offset::<ast::Impl>()?;
+
+    // If impl is not inherent then we don't really need to go any further.
+    if impl_ast.for_token().is_some() {
+        return None;
+    }
+
+    let assoc_items = impl_ast.assoc_item_list()?;
+    let first_element = assoc_items.assoc_items().next();
+    if first_element.is_none() {
+        // No reason for an assist.
+        return None;
+    }
+
+    acc.add(
+        AssistId("generate_trait_from_impl", ide_db::assists::AssistKind::Generate),
+        "Generate trait from impl".to_owned(),
+        impl_ast.syntax().text_range(),
+        |builder| {
+            let trait_items = assoc_items.clone_for_update();
+            let impl_items = assoc_items.clone_for_update();
+
+            trait_items.assoc_items().for_each(|item| {
+                strip_body(&item);
+                remove_items_visibility(&item);
+            });
+
+            syntax::ted::replace(assoc_items.clone_for_update().syntax(), impl_items.syntax());
+
+            impl_items.assoc_items().for_each(|item| {
+                remove_items_visibility(&item);
+            });
+
+            let trait_ast = ast::make::trait_(
+                false,
+                "NewTrait".to_string(),
+                HasGenericParams::generic_param_list(&impl_ast),
+                HasGenericParams::where_clause(&impl_ast),
+                trait_items,
+            );
+
+            // Change `impl Foo` to `impl NewTrait for Foo`
+            // First find the PATH_TYPE which is what Foo is.
+            let impl_name = impl_ast.self_ty().unwrap();
+            let trait_name = if let Some(genpars) = impl_ast.generic_param_list() {
+                format!("NewTrait{}", genpars.to_generic_args())
+            } else {
+                format!("NewTrait")
+            };
+
+            // // Then replace
+            builder.replace(
+                impl_name.clone().syntax().text_range(),
+                format!("{} for {}", trait_name, impl_name.to_string()),
+            );
+
+            builder.replace(
+                impl_ast.assoc_item_list().unwrap().syntax().text_range(),
+                impl_items.to_string(),
+            );
+
+            // Insert trait before TraitImpl
+            builder.insert_snippet(
+                SnippetCap::new(true).unwrap(),
+                impl_ast.syntax().text_range().start(),
+                format!("{}\n\n", trait_ast.to_string()),
+            );
```

Comment:
My first intention was to let the user fill out trait's name, possibly using a multi-cursor so that the user doesn't have to go from trait to impl to write the same name twice. But I learned that there is no multi cursor or sth similar in ra.  Anyway thanks for the correction.

Reply:
r-a does sort of have support for multi-cursor placeholders, and can be done by wrapping each of the names inside of a placeholder marker (e.g. `${0:NewTrait}`).

Wrapping names inside of a placeholder marker should be done only if a `SnippetCap` is available, but check for doing so is relatively simple (essentially just `ctx.config.snippet_cap.is_some()`).

Reply:
Thank you for the instructions. The latest commit introduces conditional use of multi-cursor.